### PR TITLE
Fixing a bug where messenger:consume could send message to wrong bus

### LINKED ISSUE #30631: [Messenger] messenger:consume - bus should be determined automatically
**Symfony version(s) affected**: 4.1+

**Description**  
The `messenger:consume` command "receives" messages from a transport and then sends them back into the bus. But if you have *multiple* busses, the command requires you to choose which bus you want. This is incorrect: the message should automatically be sent back through the correct bus.

For example:

* `MessageA` goes through `BusA` and is sent to `TransportFoo`
* `MessageB` goes through `BusB` and is *also* sent to `TransportFoo`

Now, if I run `messenger:consume` and choose `BusA`, I will receive & send both `MessageA` AND `MessageB` through `BusA`. This means that **`MessageB` will incorrectly go through BusA**.

But what about messages that are sent to the queue from an external app and *only* consumed by Messenger? In that case, I think we need to send to the default bus and/or allow the bus to be configured on a transport-by-transport basis.

**Possible Solution**  
The bus name should be attached via a stamp and read by the `Worker` to determine which bus to send to. I think the only way to do that is to pass a new BusNameMiddleware automatically into each MessageBus as the first middleware that knows the name and can add a stamp. That’s only a little weird because component users would need to know that this middleware needs to be added when creating the MessageBus. Or am I missing another way?

For messages received from an external system (no `SentStamp`), there would need to be sent to the "default" bus, or allow the bus to be configured for that transport.

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | arguably, yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30631
| License       | MIT
| Doc PR        | Not needed

This fixes #30631, where you can run `messener:consume` and accidentally sent received messages into the wrong bus.

The fix (done via middleware) is to attach a "bus name" to the `Envelope` and use it when the message is received to find that bus.

### CODE REVIEW

**Comment by weaverryan** [comment]
This is ready to go!

File: src/Symfony/Component/Messenger/RoutableMessageBus.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public function __construct(ContainerInterface $busLocator)
+    {
+        $this->busLocator = $busLocator;
+    }
+
+    public function dispatch($envelope): Envelope
+    {
+        if (!$envelope instanceof Envelope) {
+            throw new InvalidArgumentException('Messages passed to RoutableMessageBus::dispatch() must be inside an Envelope');
```
Comment (sroze) [inline src/Symfony/Component/Messenger/RoutableMessageBus.php]:
```suggestion
            throw new InvalidArgumentException('Messages passed to RoutableMessageBus::dispatch() must be inside an Envelope.');
```

**Review by sroze — APPROVED** [review]
Great, indeed. I was aiming to write such a PR with me (but this small change) 👌

**Review by chalasr — APPROVED** [review]
👍

File: src/Symfony/Component/Messenger/RoutableMessageBus.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+use Symfony\Component\Messenger\Exception\InvalidArgumentException;
+use Symfony\Component\Messenger\Stamp\BusNameStamp;
+
+/**
+ * Bus of buses that is routable using a BusNameStamp.
+ *
+ * This is useful when passed to Worker: messages received
+ * from the transport can be sent to the correct bus.
+ *
+ * @experimental in Symfony 4.2
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/RoutableMessageBus.php]: `@experimental in 4.3` (omitting `Symfony` for consistency), same for other classes added here

**Review by chalasr — CHANGES_REQUESTED** [review]

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Messenger/RoutableMessageBus.php (hunk 2 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/RoutableMessageBus.php]: Just changed all occurrences, great catch 👍

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Comment by sroze** [comment]
Merged by e512b7e but GitHub couldn't recognise it because I pushed another commit in this branch in the meantime, closing.

**Comment by raresserban** [comment]
Just a thought on this, I think it is a BC Break because if you have an application running SF 4.2 which has published a message that will then be consumed by an application running SF 4.3, you will get the error:
Envelope does not contain a BusNameStamp.

Hence I think it is a BC break. Moreover, if for example, some messages exist in a RabbitMQ queue and you updated your app to SF 4.3, then those messages can no longer be processed because they do not have this stamp. I am proposing that if the stamp is not present then the default bus to be used to handle them.

**Comment by sroze** [comment]
>  I am proposing that if the stamp is not present then the default bus to be used to handle them.

@raresserban The component being experimental, we allow BC breaks. Though, I agree with you, it doesn't cost much effort to do so. Can you submit a pull request? Thank you!

**Comment by weaverryan** [comment]
Yes, I agree. I've been trying to "keep in mind" (at least) the migration path for messages that are inside a queue at the moment of upgrading.

See symfony/symfony#31200

**Pull request closed without merging**